[miffi] iP

miffi opened the pull request:
# The Cat that Lives in your Walls :cat:
```
 |\ /|
=(O O)=
 /   \
```
Just hand your tasks to it, and it'll keep track of it in the wall. It'll respond by gestures or by handing you notes. ~~It doesn't bite.~~

## Advantages of the mascot
* Cute (a given)
* Bit of a weirdo (who lives in _walls_?)
* **Cute**

## Why
Having a cat in your vicinity immediately raises your charisma which I've made into this handy formula.
```py
def effectiveCharisma(actualCharisma: int, hasCat: int) -> int:
    if hasCat:
        return actualCharisma + 13
    else:
        return actualCharisma
```

## Features
* [x] Tasks
* [x] Deadlines
* [x] Events
* [ ] GUI
* [ ] Editing tasks

## Installation and running
1. Have a java JDK installed
1. Go to the [releases](https://github.com/miffi/ip/releases) tab
1. Get the `jar` file
1. Run the jar file (either by clicking on it or by running it from the command line with `java -jar duke.jar`)

P.S. The cat told me to also say, and I quote:
> Meow

P.S.S. The charisma joke is inspired from https://www.youtube.com/watch?v=Ny5f1BsQsPE

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

miffi committed:
Add greeting and exit messages

miffi committed:
Introduce said cat

miffi committed:
Echo input back to user

miffi committed:
Extract I/O logic into Main file

miffi committed:
Add Task with simple adding and listing

miffi committed:
Inline parseCommand

It makes more sense for the control flow to break rather than to
return whether it should end.

miffi committed:
Make bye into a command

miffi committed:
Add index information into Duke's messages

miffi committed:
Add Marking and Unmarking

miffi committed:
Add Task subclasses

miffi committed:
Typos

miffi committed:
Make the absence of a command an error

miffi committed:
Add I/O capabilities for Todos, Deadlines and Events

miffi committed:
refactor: move component handling entirely into Duke

miffi committed:
refactor: Move task creation logic into a factory method for Task

miffi committed:
Make Task an abstract class

It is not going to be called by itself. It needs a category.

miffi committed:
refactor: Change Main to Duke and Duke to TaskList

Some scripts are dependent on the main entrypoint being Duke

miffi committed:
Add Automated Text UI Testing

miffi committed:
Add deletion capabilities

miffi committed:
refactor: Make TaskNotFound synthesize message from index

miffi committed:
chore: formatting

miffi committed:
Add state saving and loading

This uses java's built in serialization as it's the easiest and most
fault tolerant way to save the entire state.

miffi committed:
Merge branch 'branch-Level-7'

miffi committed:
refactor: remove redundant `this`

miffi committed:
refactor: Split code to packages

miffi committed:
Use DateTime instead of strings for dates in Tasks

miffi committed:
Merge branch 'branch-Level-8'

miffi committed:
Abstract storage to a separate class

miffi committed:
Add final to methods that shouldn't be overridden

miffi committed:
Abstract Ui out into a separate class

miffi committed:
Remove dead code

miffi committed:
Move task component validation to subclasses

miffi committed:
Abstract commands into their respective classes

miffi committed:
Make storage in Duke final

miffi committed:
Merge branch 'more-oop'

miffi committed:
Merge branch 'add-gradle-support'

miffi committed:
Change class name in gradle

I don't have the same directory structure as they expect.

miffi committed:
refactor: Move Parser to top level package

Mistakenly created it in the Task package.

miffi committed:
Fix and test edge cases of the parser

miffi committed:
Handle component validation edge case.

miffi committed:
Remove ending newline from taskList toString

miffi committed:
Add more Task test cases

miffi committed:
Add docs

miffi committed:
Follow coding standard

miffi committed:
Remove redundant parameter documentation

Looking at the coding standard, I realised that a lot of the
parameter documentation isn't necessary as they are obvious.

miffi committed:
fix: fixes list output regression

The change to the list output broke when there was zero elements.
It also showed the first element twice.

miffi committed:
Add find command

miffi committed:
Merge branch 'branch-Level-9'

miffi committed:
Merge branch 'branch-A-JavaDoc'

miffi committed:
Merge branch 'branch-A-CodingStandard'

miffi committed:
Fix extra space at the start of task description

miffi committed:
Handle invalid date parameter

Previously it just unceremoniously crashed.

miffi committed:
Add credit for inspiration

miffi committed:
style: use expression lambdas

miffi committed:
Add test cases for valid Task components

miffi committed:
Add checkstyle configuration

miffi committed:
style: fix checkstyle warnings

miffi committed:
Merge branch 'branch-A-CheckStyle'

miffi committed:
Fix NullPointerException in empty component

There was a check to make sure that when no string without a / comes
after a component, it caused an exception. It did this by checking the
length, but the `StringBuilder` is null until after getting the first
part of the component parameter. Therefore, the length check caused a
NullPointerException.

The fix was to check if the builder is `null` instead.

miffi committed:
Add preliminary GUI

miffi committed:
refactor: move GUI to a separate ui package

miffi committed:
Save data on GUI close

miffi committed:
Add missing javaDoc

miffi committed:
Hook Duke to GUI

miffi committed:
Change main class to Launcher

miffi committed:
Enable assertions in gradle

miffi committed:
Add null safety asserts

The code has too many implicit assumptions about null safety. The
program can crash from a null pointer in some obscure place from a
regression.

The assumptions are made explicit with non-null assertions to fix this.

Making all the non-null assumptions explicit stops null parameters at
the earliest possible point of execution. This will hopefully make the
error show nearer to the source of the null, so that they can be more
easily debugged.

miffi committed:
Replace for loops with streams

miffi committed:
refactor: Extract methods and variables

A lot of the functions use a single level of abstraction. Therefore, the
code is hard to understand.

The fix was to extract out the mthods and variables to make the code
streamlined.

miffi committed:
refactor: remove magic strings in the parser

The parser uses strings to match commands. This makes checking whether
the commands are valid to be error prone as a typo would mess it up. It
also makes it very inflexible as the string can only match one type.

To solution was to make an enum of the command type and moving command
type parsing logic to a static method in the enum.

This decouples checking the parsing to the enum, and it opens up to
more flexible parsing. This also means that we can write the parser and
be sure of exhaustiveness.

miffi committed:
refactor: extract more variables and methods

miffi committed:
refactor: remove dead code

miffi committed:
refactor: extract parser empty builder check to method

miffi committed:
refactor: remove dead exit check code

The code is not being used any more so it's time to remove it.

`ExitCommand` can still be used for prompting during exit or something
so that should remain.

miffi committed:
Merge pull request #1 from miffi/branch-A-CodeQuality

Branch for code quality

miffi committed:
Merge branch 'master' into branch-A-Assertions

miffi committed:
Merge pull request #2 from miffi/branch-A-Assertions

Branch of assertions

miffi committed:
Merge pull request #3 from miffi/branch-A-Streams

Replace for loops with streams

miffi committed:
style: fix checkstyle errors

miffi committed:
Add Gradle CI

miffi committed:
Remove redirection test

miffi committed:
Rename duke to cat

miffi committed:
Fix description not being detected

miffi committed:
Add a help command

miffi committed:
Handle multiple indices in completion and deletion

myang2020 commented on src/main/java/cat/Cat.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Saves the state to storage.
+     */
+    public void save() {
+        try {
+            storage.writeTaskList(tasks);
+        } catch (FileNotFoundException e) {
+            System.out.println("Cannot find state file \"" + FILE_NAME + "\"");
+        } catch (IOException e) {
+            System.out.println("Cannot write to state file \"" + FILE_NAME + "\"");
+        }
+    }
```
Great that you separated a save() function from main function. Could improve even more by putting save() in a storage class.

myang2020 commented on src/main/java/cat/Parser.java:
```diff
@@ -0,0 +1,221 @@
+public class Parser {
+    private static final String DESCRIPTION_IDENTIFIER = "description";
+
+    private enum CommandType {
+        Todo,
+        Deadline,
+        Event,
+        Mark,
+        Unmark,
+        Delete,
+        Find,
+        List,
+        Help,
+        Bye;
```
Great use of enums to denote a valid set of commands!

myang2020 commented on src/main/java/cat/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     * can either be that no parameters are given, or that
+     * a parameter is missing.
+     */
+    public static class InvalidCommandData extends Exception {
+        public InvalidCommandData() {
+            super("Parameters to command not given.");
+        }
+
+        public InvalidCommandData(String key) {
+            super("No value given to " + key);
+        }
+    }
```
Perhaps create a custom DukeException class

myang2020 reviewed (commented):
Sorry, reviewed wrong person.

0nesheep commented on src/main/java/cat/Parser.java:
```diff
@@ -0,0 +1,221 @@
+        String input = getCommandInput(scanner);
+        var components = parseComponents(input);
+
+        switch (type) {
+        case Todo:
+            return new AddCommand(AddCommand.Type.Todo, components.description, components.parameters);
+        case Deadline:
+            return new AddCommand(AddCommand.Type.Deadline, components.description, components.parameters);
+        case Event:
+            return new AddCommand(AddCommand.Type.Event, components.description, components.parameters);
+        case Mark:
+            return new CompleteCommand(parseIndices(input), true);
+        case Unmark:
+            return new CompleteCommand(parseIndices(input), false);
+        case Delete:
+            return new DeleteCommand(parseIndices(input));
+        case Find:
+            return new FindCommand(input);
+        default:
+            throw new IllegalStateException("Unexpected type: " + type);
+        }
+    }
```
I really like how you handled the command inputs! It's so clean 👍

0nesheep commented on src/main/java/cat/Parser.java:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    /**
+     * Exception when the command is unknown.
+     */
+    public static class InvalidCommandType extends Exception {
+        private final String command;
+
+        /**
+         * Creates an exception about an unknown command string.
+         *
+         * @param command The name of the unknown command
+         */
+        public InvalidCommandType(String command) {
+            super("Command \"" + command + "\" is invalid or not yet implemented.");
+            this.command = command;
+        }
+
+        public String getCommand() {
+            return command;
+        }
+    }
+
+    /**
+     * Exception when the data given to a particular
+     * command doesn't match what is expected. The problem
+     * can either be that no parameters are given, or that
+     * a parameter is missing.
+     */
+    public static class InvalidCommandData extends Exception {
+        public InvalidCommandData() {
+            super("Parameters to command not given.");
+        }
+
+        public InvalidCommandData(String key) {
+            super("No value given to " + key);
+        }
+    }
```
Might it be better to extract these exceptions to their own separate classes? Or at least a CatException class?

miffi replied:
Hmm I think that makes sense. I was making the exceptions near where they are thrown from, but it might be better to centralise them. Well I'm still not 100% sold on that, but I think having a CatException to indicate that this is for internal errors is something I'll absolutely implement. Thanks!

0nesheep commented on src/main/java/cat/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+                .collect(Collectors.joining()).stripTrailing();
+    }
+
+    /**
+     * An exception used to signal that a task from a given request doesn't exist.
+     */
+    public static class TaskNotFound extends Exception {
+        public TaskNotFound(int index, Throwable cause) {
+            super("Could not find task " + (index + 1), cause);
+        }
+    }
+}
```
Similar to other comment on exceptions.

0nesheep commented on src/main/java/cat/command/AddCommand.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Constructs a command that adds a task to the task list.
+     *
+     * @param type       the type of the task
+     * @param components the components parsed from the user command
```
I think we need to capitalize the first letter and end the statement with a ```.```? Similar comment applies to other Javadoc, will not repeat to avoid too much repetition!

miffi replied:
I don't know how I missed that guideline. Thanks for pointing that out. ^^

0nesheep reviewed (commented):
LGTM! Your error handling is very well done! Just some minor nit on extracting exception classes and a minor Javadoc styling error.

Also the cat is so cute <3 but get it out of my wall please :(

miffi committed:
Add response classes and UI

Right now, as the week 6 better gui incentive outlined, the chat like Ui
is not really that indicative of the interaction between the app and the
user.

Therefore, I'm refactoring some of the code to make the UI emit a
Response class for the app, and this respose will seem very different
from the user's input.

miffi committed:
Finish UI

miffi committed:
Make ErrorResponse able to construct from string

Some of the code makes Exceptions with a message string just to pass
them to ErrorResponse. This indirection makes no sense.

Therefore, this commit adds an overloaded constructor to construct from
a String as well.

miffi committed:
Fix checkstyle errors

miffi committed:
Merge branch 'branch-A-BetterGUI'

miffi committed:
Add handling of invalid date time range

miffi committed:
Use getMessage instead of toString for errors

miffi committed:
Remove redundant Pane in Response

miffi committed:
Remove invalid help and images

miffi committed:
Fix style warnings from SoC SE Bot

miffi committed:
Fix javadoc style errors

miffi committed:
Add user guide

miffi committed:
Remove redundant Command storage parameter

miffi committed:
Fix branding
